File: CompetitiveProgramming/LinkedList/ReverseLinkedList/reverseLinkedList.cpp
//https://leetcode.com/problems/reverse-linked-list/

public class ListNode{
    int val 
    ListNode* next;


    ListNode(){}

    ListNode(int val){
        this.val = val;
    }

    ListNode(int val , ListNode* nextNode){
        this.val = val;
        this.next = nextNode;
    }
}


class Solution{

    //Time Complexity : O(n)
    //Space Complexity : O(1)

    public: 

        ListNode* reverseList(ListNode* head){

            ListNode* prevNode = NULL;
            ListNode* currentNode = head;
            ListNode* nextNode;


            while(currentNode){

                nextNode = currentNode->next;
                currentNode->next = prevNode;

                prevNode = currentNode;
                currentNode = nextNode;

            }

            return prevNode;
        }

        ListNode reverseList2(ListNode* head){

                ListNode* updatedHead = NULL;

                while(head){
                    ListNode* next = head->next;
                    head->next = updatedHead;

                    updatedHead = head;
                    head = next;
                }

                return updatedHead;


        }

    
};

File: CompetitiveProgramming/LinkedList/ReverseLinkedList/reverseLinkedList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


class Solution:

    # Time Complexity : O(n)
    # Space Complexity : O(1)
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:

        updatedHead = None
        while(head):

            next = head.next
            head.next = updatedHead

            updatedHead = head
            head = next

        return updatedHead


